boards/esp32-ethernet-kit/Kconfig

```
@@ -0,0 +1,21 @@
+# Copyright (c) 2020 HAW Hamburg
+#
+# This file is subject to the terms and conditions of the GNU Lesser
+# General Public License v2.1. See the file LICENSE in the top level
+# directory for more details.
+
+config BOARD
+    default "esp32-ethernet-kit" if BOARD_ESP32_ETHERNET_KIT
+
+config BOARD_ESP32_ETHERNET_KIT
+    bool
+    default y
+    select BOARD_COMMON_ESP32
+    select CPU_MODEL_ESP32_WROVER
```

The board uses the ESP-WROVER-B module. We have to introduce this as a new CPU model variant.
```suggestion
    select CPU_MODEL_ESP32_WROVER_B
```
To get this working, you have to add
```
 config CPU_MODEL_ESP32_WROVER_B
     bool
     select CPU_FAM_ESP32
```
in `cpu/esp32/Kconfig` here https://github.com/RIOT-OS/RIOT/blob/892370121ded21733f2a10b35563646f2472cc74/cpu/esp32/Kconfig#L27-L30 with

---

I have the v1.2 board (can't find it mentioned on the ESP site) - it has the ESP32-WROVER-E mounted.

Should we care about revisions of the board?

---

Hm ... :thinking: @leandrolanzieri Here we run first time into the problem using the modul as CPU model. Different versions of the board use different ESP32-WROVER-* modules. How should we deal with it?

---

Having modules as `CPU_MODEL` seems like a strange concept.
What is the difference between the module versions anyway?

---

> Having modules as `CPU_MODEL` seems like a strange concept.
> What is the difference between the module versions anyway?

Indeed. That was also my first thought when it was introduced with PR #14423. Therefore we had a long discussion there, especially  in comments https://github.com/RIOT-OS/RIOT/pull/14223#issuecomment-651011698, https://github.com/RIOT-OS/RIOT/pull/14223#issuecomment-651672663, https://github.com/RIOT-OS/RIOT/pull/14223#issuecomment-651700110 and https://github.com/RIOT-OS/RIOT/pull/14223#issuecomment-652382494.

All modules use the ESP32-D0WD CPU and have at least 4 MByte flash. Some of them can also have 8 or 16 MByte flash. ESP32-WROVER modules already have 4 or 8 MByte of SPI-RAM on module. In fact, it isn't possible at the moment to determine the size of the RAM or the flash from the CPU model. But finally I accepted to use modules as CPU model because it makes it easier to determine from the configuration what CPU or module a board uses.

---

> I have the v1.2 board (can't find it mentioned on the ESP site) - it has the ESP32-WROVER-E mounted.

Are there other differences between v1.0, v1.1, v1.2? One difference between v1.0 and v1.1 I found is that in v1.0 GPIO4 was used for the JTAG TRST signal which can be switched off.

---

I filed a bug asking for an update: https://github.com/espressif/esp-idf/issues/5572

Both GPIO2 and GPIO4 are busy on rev 1.0.

---

> I filed a bug asking for an update: [espressif/esp-idf#5572](https://github.com/espressif/esp-idf/issues/5572)
> 
> Both GPIO2 and GPIO4 are busy on rev 1.0.

Yes, but both can be switched off. As I said, JTAG TRST signal isn't required and can be switched off.

---

I would suggest to introduce a further pseudomodule in `boards/esp32-ethernet-kit/Makefile.include` which allows to specify version 1.0. We did it in same way for the `esp32-ttg-t-beam` board.
```
PSEUDOMODULE += esp32_ethernet_kit_v1_0
```
Then you could change the definition of `PWM_CHANNELS` as following.
```diff
- #define PWM0_GPIOS  { GPIO2, GPIO4 }
+ #ifdef MODULE_ESP32_ETHERNET_KIT_V1_0
+ #define PWM0_GPIOS  { GPIO4 }
+ #else
+ #define PWM0_GPIOS  { GPIO2, GPIO4 }
+ #endif
```
Of course, it would have to be documented.

---

Sounds good, added it.

---

Well the discussion focused on trying to be more specific than just saying `CPU_MODEL=esp32`. If we say that the modules should not be considered `CPU_MODEL`s then a module would be strictly speaking a board (CPU + external connections to Flash or RAM). If we go down this path we can think as having the single `CPU_MODEL=esp32-d0wd` and then common `BOARD_COMMON_ESP32_WROVER_B` symbols for the modules. All this considering it actually makes sense to know which module is being used...

---

Is there consensus here on some specific change I should do?

---

ESP has updated the docs now for the v1.2 board: https://docs.espressif.com/projects/esp-idf/en/latest/esp32/hw-reference/esp32/get-started-ethernet-kit.html

Seems like nothing interesting other than the WROVER module change happened.

---

According to the discussion in issue #14681, we will need a separate board directory for each revision. So I wonder whether we have to restructure this PR.

If I'm right, we have three revisions of the same board on the market v1.0, v1.1 and v1.2. AFAIK, the only differences are that

- GPIO2 is not available in v1.0 and
- v1.0/v1.1 use WROVER_B while v1.2 uses WROVER_E.

Right? Since we do not distinguish the WROVER modules according to the discussion in PR #14223, the only difference is the availability of GPIO4.

There are two options to have a separate directory for each revision:

1. Create a common board definition `esp32-ethernet-kit` in `boards/common` and then three board definitions `esp32-ethernet-kit-v1-0`, ....
2. Define a base revision `esp32-ethernet-kit-v1-0` that makes all conditional definitions, and create two additional board definitions `esp32-ethernet-kit-v1-1` and `esp32-ethernet-kit-v1-2` that only enable the corresponding pseudo-module and include everything else from the base revision (This approach was used by @benpicco for `openlabs-kw41z-mini-256kib`)

Option 1 seems to be an overkill because of this small difference.

---

While trying to fix this, I ran into some hard-coded includes, like in RIOT/cpu/esp32/vendor/esp-idf/wpa_supplicant/port/Makefile:

```
INCLUDES += -I$(BOARDDIR)/include
```
Should I symlink the include folders between them? It does not work doing like in `openlabs-kw41z-mini-256kib` (`-I$(RIOTBOARD)/otherboard/include`)

If I add include of the board Makefile.include there as well, it still fails since the board include folder doesn't exist.

---

Hm, good point. We already use symlinks in RIOT. So I would say that it should be fine to create a symlink. We include everything from the base revision, why not to link to the include directory. However, I would like to ask for the opinion of @benpicco first.

Another solution would be to resolve the include dependency. With the following changes 
- `cpu/esp32/include/sdk_conf.h`
   ```diff
   - #include "board.h"
   ```
- `cpu/esp32/vendor/esp-idf/include/esp32/esp_system.h`
   ```diff
   - #include "esp_sleep.h"
   ```
you could remove this dependency in all `wpa_supplicant` makefiles.
```
cpu/esp32/vendor/esp-idf/wpa_supplicant/port/Makefile
cpu/esp32/vendor/esp-idf/wpa_supplicant/src/wpa2/utils/Makefile
cpu/esp32/vendor/esp-idf/wpa_supplicant/src/wpa2/eap_peer/Makefile
cpu/esp32/vendor/esp-idf/wpa_supplicant/src/wpa2/tls/Makefile
cpu/esp32/vendor/esp-idf/wpa_supplicant/src/crypto/Makefile
```

---

If the dependency is not needed in the first place, that would of course be the easiest solution.

---

> If the dependency is not needed in the first place, that would of course be the easiest solution.

I checked it for `esp_wifi_enterprise` which is the configuration that uses all the makefiles in the vendor code with this dependecy.

Just for information: Setting `-I$(BOARDDIR)/include` as include path was necessary in these makefiles because the `INCLUDE` variable as provided by the make system had to be overriden due to the search order of the header files. For all other makefiles in the vendor code the search order as provided by the make system is fine.

---

Well if `wpa_supplicant` doesn't need the `$(BOARDDIR)/include`, I'd be happy to remove it - want to open a separate PR?

---

@benpicco Done

---

And merged - so this issue should be resolved after a rebase.

---

I split it to three separate boards (cherrypicked the wpa_supplicant change to avoid force push).
The documentation now has common group and then groups for each version with more details.

---

@gschorcht Any comments on the latest code?